### ISSUE TITLE
Unable to download efcarson_aoi_1

### ISSUE THREAD

**Issue by lbross** [PSU-CSAR/django-ebagis#30]
Low priority but can you take a look? When I try to download this aoi using PC-BAGIS, I get a .zip file that I cannot open. No errors from server AFAIK.

**Comment by jkeifer** [comment]
I downloaded the file generated from one of your requests and I was able to open it just fine. I also looked at the zipfiles on the server and they all seemed fine. Can you send me the zip that wouldn't open?

**Comment by lbross** [comment]
Too big to post to GigHub or attach to an e-mail. I uploaded it to basins: /BAGIS/For_Jarrett/efcarson_aoi_1.zip

**Comment by jkeifer** [comment]
Great, I will take a look later today and let you know what I can determine.

**Comment by jkeifer** [comment]
I'm sorry for the delay. I took a look at the file you uploaded and saw the problem right away: it was only part of the zipfile. The real zipfile is 49MB, but what you uploaded was only 39MB.

I downloaded the zipfile via a web browser a couple times and each time got the whole thing. It may be a problem with files past a certain size and the PC_BAGIS downloader, or perhaps it is an issue with the way the file is being served that causes problems knowing when the file is complete.

Can you take another look at the downloader code to see if you can find anything obvious in light of the fact that your only got part of the file? I will also try to work on serving the files in a different way, such that the server can send along the size of the file at the outset so you know how big a file to expect.

**Comment by jkeifer** [comment]
I have added some extra info along with the files that I am serving. You should now get the `Content-Length` parameter in the HTTP header for the response. Also, I didn't test it, so you may uncover bugs, but you should be able to request just a byte range of the file by putting the start and stop bytes in the request header under the `range` parameter, like `range: 1500-45673130` to get from byte 1500 to byte 45673130. I still need to rework the parser for that request to allow open-ended ranges, such as just specifying a start or just and end byte. But assuming it works, this should allow download-resume functionality, for when a download is interrupted.

These changes are already available on the current testing instance, so you don't have to wait for me to deploy my new code to start testing this. I know I am a bit behind on getting my changes deployed; I have not had as much time to devote to the project as I had hoped. I am getting really close though, so hopefully they'll get deployed in the next couple days.

Some notes for my reference:

The new download implementation primarily came from [stackoverflow/Resumable downloads when using PHP to send the file?](http://stackoverflow.com/questions/157318/resumable-downloads-when-using-php-to-send-the-file). I also tried to use the suggestions from [stackoverflow/serving-large-files-with-high-loads-in-django](http://stackoverflow.com/questions/8600843/serving-large-files-with-high-loads-in-django), but ran into problems and decided not to spend anymore time on this issue right now. I believe the current solution should read the file in pieces for normal downloads, but for ranged requests, the whole chunk will be read into memory on the server, which is not an ideal solution, particularly if we ever have high amounts of download traffic.

However, it is important to note that the current solution for serving downloads will not scale, and django is not designed for this use-case. Instead, if download traffic becomes high, then it would be best to move  to a solution where the files are not served by django, but the reverse proxy itself, so django doesn't get tied up by any filestreams. One thing to maybe checkout for this is [django-downloadview](http://django-downloadview.readthedocs.org/en/latest/overview.html), which seems to pass the file reference to the reverse proxy for handling, though I don't know if this it compatible with IIS.

**Comment by lbross** [comment]
Looks like the content-type for the zip file has changed from application/zip to application/x-zip-compressed which broke my code but easy to fix. Will this be the content-type for downloads going forward?

**Comment by lbross** [comment]
Your changes appear to have fixed the problem. I was just able to successfully download this AOI once I changed the content-type.

**Comment by jkeifer** [comment]
Just to respond to the question about content-types: I switched the content-type from a hard-coded value of `application/zip` to using the python mimetypes module so the value will be automatically determined based on the file itself.

I believe using the module is a more robust solution, allows server-side support for more filetypes in the future, and is generally a better way to handle the content-types. However, if the change to this module causes problems in the future, e.g., unexpected or changing content-types returned, then we can reevaluate it's use.

I'm glad the download is working now. As you are able to download the file this issue should remain closed, but I did have some outstanding work to do regarding requested byte ranges. I will create a new issue for that work.
